File: Service/Service.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Grpc.Core;
using Shared;

namespace Service
{


    public class MyService : Svc.SvcBase
    {
        public override Task<CalculateReply> Calculate(CalculateRequest request, ServerCallContext context)
        {
            long result = -1;
            switch (request.Op)
            {
                case "+":
                    result = request.X + request.Y;
                    break;
                case "-":
                    result = request.X - request.Y;
                    break;
                case "*":
                    result = request.X * request.Y;
                    break;
                case "/":
                    if (request.Y != 0)
                    {
                        result = (long)request.X / request.Y;
                    }
                    break;
                default:
                    break;
            }
            return Task.FromResult(new CalculateReply { Result = result });
        }

        public override async Task Median(IAsyncStreamReader<Temperature> requestStream, IServerStreamWriter<Temperature> responseStream, ServerCallContext context)
        {
            Console.WriteLine("Median");
            var vals = new List<double>();
            while (await requestStream.MoveNext())
            {
                var temp = requestStream.Current;
                vals.Add(temp.Value);
                double med = 0;
                if (vals.Count == 10)
                {
                    var arr = vals.ToArray();
                    Array.Sort(arr);
                    med = (arr[4] + arr[5]) / 2;
                    vals.Clear();
                    await responseStream.WriteAsync(new Temperature { Timestamp = temp.Timestamp, Value = med });
                }
            }
        }

        static void Main(string[] args)
        {
            int port = int.Parse(args[0]);
            var pair = new KeyCertificatePair(
                File.ReadAllText("cert/service.pem"),
                File.ReadAllText("cert/service-key.pem")
                );
            var creds = new SslServerCredentials(new[] { pair });
            var server = new Server
            {
                Services = { Svc.BindService(new MyService()) },
                Ports = { new ServerPort("0.0.0.0", port, creds) }
            };
            server.Start();
            Console.WriteLine($"Server listening at port {port}. Press any key to terminate");
            Console.ReadKey();
        }
    }
}

File: python/service.py
from concurrent import futures
import statistics
import sys

import grpc

import contracts_pb2
import contracts_pb2_grpc


class Service(contracts_pb2_grpc.SvcServicer):
    def Calculate(self, request, context):  # noqa
        result = -1
        if request.op == "+":
            result = request.x + request.y
        elif request.op == "-":
            result = request.x - request.y
        elif request.op == "*":
            result = request.x * request.y
        elif request.op == "/":
            if request.y != 0:
                result = request.x // request.y
        return contracts_pb2.CalculateReply(result=result)

    def Median(self, request_iterator, context):  # noqa
        vals = []
        for temp in request_iterator:
            vals.append(temp.value)
            med = 0
            if len(vals) == 10:
                med = statistics.median(vals)
                vals = []
                yield contracts_pb2.Temperature(timestamp=temp.timestamp, value=med)


with open("../cert/service-key.pem", "rb") as f:
    key = f.read()

with open("../cert/service.pem", "rb") as f:
    cert = f.read()

creds = grpc.ssl_server_credentials(
    [
        (
            key,
            cert,
        ),
    ]
)

port = int(sys.argv[1])
svc = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
contracts_pb2_grpc.add_SvcServicer_to_server(Service(), svc)
svc.add_secure_port(f"[::]:{port}", creds)
svc.start()
print(f"service listening on port {port}...")
svc.wait_for_termination()
